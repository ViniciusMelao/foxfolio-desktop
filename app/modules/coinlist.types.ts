export interface Coinlist {
  [symbol: string]: CoinlistEntry;
}

export interface CoinlistEntry {
  FullName: string;
  ImageUrl: string;
  SortOrder: string;
}
