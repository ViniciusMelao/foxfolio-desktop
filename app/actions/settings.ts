import { CurrencyFocus, SettingsType } from '../modules/settings.types';
import { getSettings } from '../selectors/selectGlobalState';
import { Action, Dispatch, GetState, ThunkAction } from './actions.types';
import { requestTickerUpdate } from './ticker';

function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}

export function setPortfolioFocus(currencyFocus: CurrencyFocus): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(saveSettings({ ...getSettings(getState()), currencyFocus }));
  };
}

export function saveSettingsAndUpdateTicker(settings: SettingsType): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(saveSettings(settings));
    dispatch(requestTickerUpdate([], settings.fiatCurrency, settings.cryptoCurrency));
  };
}
