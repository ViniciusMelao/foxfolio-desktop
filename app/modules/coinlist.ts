import { Action } from '../actions/actions.types';
import { Coinlist } from './coinlist.types';

export const coinlist = (state: Coinlist = {}, action: Action): Coinlist => {
  switch (action.type) {
    case 'RECEIVE_COIN_LIST':
      return action.coinlist;
    default:
      return state;
  }
};
