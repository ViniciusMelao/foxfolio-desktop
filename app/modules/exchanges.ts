import _ from 'lodash';
import { Action } from '../actions/actions.types';
import { generateId } from '../helpers/reducers';
import {
  AddExchangeAction,
  DeleteExchangeAction,
  Exchange,
  ExchangeInstanceActions,
  Exchanges,
  ExchangeTypeKeys,
  FailedExchangeRequestAction,
  UpdateExchangeBalancesAction,
  UpdateExchangeCredentialsAction,
  UpdateExchangeLedgerAction,
  UpdateExchangeTradesAction,
} from './exchanges.types';

export function exchanges(state: Exchanges = {}, action: Action): Exchanges {
  switch (action.type) {
    case ExchangeTypeKeys.ADD_EXCHANGE:
      return addExchange(state, action);
    case ExchangeTypeKeys.DELETE_EXCHANGE:
      return deleteExchange(state, action);
    case ExchangeTypeKeys.UPDATE_EXCHANGE_CREDENTIALS:
      return reduceInstance(state, action)(updateExchangeCredentials);
    case ExchangeTypeKeys.UPDATE_EXCHANGE_BALANCES:
      return reduceInstance(state, action)(updateExchangeBalances);
    case ExchangeTypeKeys.UPDATE_EXCHANGE_LEDGER:
      return reduceInstance(state, action)(updateExchangeLedger);
    case ExchangeTypeKeys.UPDATE_EXCHANGE_TRADES:
      return reduceInstance(state, action)(updateExchangeTrades);
    case ExchangeTypeKeys.INCREMENT_EXCHANGE_REQUEST_COUNTER:
      return reduceInstance(state, action)(incrementExchangeRequestCounter);
    case ExchangeTypeKeys.FAILED_EXCHANGE_REQUEST:
      return reduceInstance(state, action)(failedRequest);
    default:
      return state;
  }
}

function addExchange(state: Exchanges, action: AddExchangeAction): Exchanges {
  const newExchange: Exchange = {
    id: generateId(Object.keys(state)),
    type: action.exchangeType,
    credentials: action.credentials,
    balances: {},
    ledger: [],
    trades: [],
  };
  return { ...state, [newExchange.id]: newExchange };
}

function deleteExchange(state: Exchanges, action: DeleteExchangeAction): Exchanges {
  return _.omit(state, action.id);
}

function updateExchangeCredentials(
  state: Exchange,
  action: UpdateExchangeCredentialsAction
): Exchange {
  return assign(state, {
    credentials: action.credentials,
  });
}

function updateExchangeBalances(state: Exchange, action: UpdateExchangeBalancesAction): Exchange {
  return assign(state, noError, decrementOpenRequests(state), {
    balances: action.balances,
  });
}

function updateExchangeLedger(state: Exchange, action: UpdateExchangeLedgerAction): Exchange {
  return assign(state, noError, decrementOpenRequests(state), {
    ledger: mergeArraysById(state.ledger, action.ledger),
  });
}

function updateExchangeTrades(state: Exchange, action: UpdateExchangeTradesAction): Exchange {
  return assign(state, noError, decrementOpenRequests(state), {
    trades: mergeArraysById(state.trades, action.trades),
  });
}

function incrementExchangeRequestCounter(state: Exchange): Exchange {
  return assign(state, {
    openRequests: (state.openRequests ? state.openRequests : 0) + 1,
  });
}

function failedRequest(state: Exchange, action: FailedExchangeRequestAction): Exchange {
  return assign(state, decrementOpenRequests(state), {
    error: action.error,
  });
}

const noError: Partial<Exchange> = {
  error: undefined,
};

const decrementOpenRequests: ((state: Exchange) => Partial<Exchange>) = (state: Exchange) => ({
  openRequests: (state.openRequests ? state.openRequests : 0) - 1,
});

type InstanceReducer<A extends ExchangeInstanceActions> = (state: Exchange, action: A) => Exchange;

function reduceInstance<A extends ExchangeInstanceActions>(
  state: Exchanges,
  action: A
): (reducer: InstanceReducer<A>) => Exchanges {
  return instanceReducer => ({ ...state, [action.id]: instanceReducer(state[action.id], action) });
}

function mergeArraysById<T extends { id: any }>(array: T[], newArray: T[]): T[] {
  return _.unionWith(array, newArray, (a, b) => _.isEqualWith(a, b, _.property('id')));
}

const assign: <T>(a: T, ...b: Array<Partial<T>>) => T = (a, ...b) => Object.assign({}, a, ...b);
