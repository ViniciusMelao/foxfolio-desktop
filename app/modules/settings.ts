import { Action } from '../actions/actions.types';
import { SettingsType } from './settings.types';

export const MINIMUM_BALANCE = 0.05;

export const initialSettings: SettingsType = {
  fiatCurrency: 'USD',
  cryptoCurrency: 'BTC',
  hideZeroBalances: false,
  includeFiat: true,
  currencyFocus: 'fiat',
  theme: 'light',
};

export default function settings(
  state: SettingsType = initialSettings,
  action: Action
): SettingsType {
  switch (action.type) {
    case 'SAVE_SETTINGS':
      return action.settings;
    default:
      return state;
  }
}
