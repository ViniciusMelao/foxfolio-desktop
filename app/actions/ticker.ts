import 'whatwg-fetch'; // Has to be imported before ccxt

import * as ccxt from 'ccxt';
import dayjs from 'dayjs';
import _ from 'lodash';
import { getHistoryEntry, getTickerEntry } from '../helpers/ticker';
import { Coinlist } from '../modules/coinlist.types';
import { Exchange, Exchanges } from '../modules/exchanges.types';
import { HistoryData, Ticker, TickerEntry } from '../modules/ticker.types';
import { Wallet } from '../modules/wallets.types';
import { getExchanges, getHistory } from '../selectors/selectGlobalState';
import { Action, Dispatch, GetState, ThunkAction } from './actions.types';
import startTimer from './timer';

const REFRESH_TIME_IN_MS = 30000;

function fetchingTickerUpdate(): Action {
  return {
    type: 'LAST_UPDATED',
    key: 'ticker',
    time: new Date(),
  };
}

function receiveTickerUpdate(rawTicker: RawTicker): Action {
  return {
    type: 'TICKER_UPDATE',
    ticker: formatRawTicker(rawTicker),
  };
}

function receiveHistory(fsym: string, tsym: string, history: HistoryData): Action {
  return {
    type: 'HISTORY_UPDATE',
    fsym,
    tsym,
    history,
  };
}

function receiveCoinList(coinlist: Coinlist): Action {
  return {
    type: 'RECEIVE_COIN_LIST',
    coinlist,
  };
}

export function requestTickerUpdate(
  extraSymbols: string[] = [],
  fiatCurrency?: string,
  cryptoCurrency?: string
): ThunkAction {
  return async (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchingTickerUpdate());

    const state = getState();
    fiatCurrency = fiatCurrency || state.settings.fiatCurrency;
    cryptoCurrency = cryptoCurrency || state.settings.cryptoCurrency;

    const symbols = getSymbolsFromTransactions(
      state.exchanges,
      state.wallets,
      fiatCurrency,
      cryptoCurrency,
      extraSymbols
    );
    try {
      let ticker = await getTickerForSymbols(symbols);
      ticker = calculateFiatValues(
        ticker,
        fiatCurrency,
        cryptoCurrency,
        getTickerEntry(ticker, cryptoCurrency, fiatCurrency)
      );
      ticker = await requestMissingTickerUpdateFromExchange(ticker, getExchanges(state), symbols);
      dispatch(receiveTickerUpdate(ticker));
    } catch (error) {
      console.error(error); // TODO Show this error to the user? Logging?
    }
  };
}

const getTickerForSymbols = (symbols: Symbols) => {
  const fsyms = symbols.from.join(',');
  const tsyms = symbols.to.join(',');
  if (fsyms && tsyms) {
    return fetch(
      `https://min-api.cryptocompare.com/data/pricemultifull?fsyms=${fsyms}&tsyms=${tsyms}`
    )
      .then(result => result.json())
      .then(result => result.RAW);
  }
  return Promise.reject('No symbols');
};

const calculateChange = (tickerEntry: any, tickerEntryFiat: any): number => {
  const open24hour = tickerEntry.OPEN24HOUR * tickerEntryFiat.OPEN24HOUR;
  const currentPrice = tickerEntry.PRICE * tickerEntryFiat.PRICE;

  return (currentPrice - open24hour) / open24hour * 100;
};

const calculateFiatValues = (
  ticker: Ticker,
  fiatCurrency: string,
  cryptoCurrency: string,
  tickerEntryFiatCrypto: TickerEntry
): Ticker => {
  return _.mapValues(ticker, (entry, asset) => {
    if (asset === fiatCurrency || asset === cryptoCurrency) {
      return entry;
    }
    return {
      ...entry,
      [fiatCurrency]: {
        ...(entry ? entry[fiatCurrency] : {}),
        CHANGEPCT24HOUR: calculateChange(
          getTickerEntry(ticker, asset, cryptoCurrency),
          tickerEntryFiatCrypto
        ),
      },
    };
  });
};

const requestMissingTickerUpdateFromExchange = async (
  ticker: Ticker,
  exchanges: Exchanges,
  symbols: Symbols
) => {
  const updatedTicker: Ticker = { ...ticker };

  try {
    for (const fsym of symbols.from) {
      if (!ticker[fsym]) {
        for (const exchange of _.values(exchanges)) {
          if (exchangeHasSymbol(fsym, exchange)) {
            const connector: ccxt.Exchange = new ccxt[exchange.type](exchange.credentials);
            const markets: {
              [symbol: string]: ccxt.Market;
            } = (await connector.loadMarkets()) as any;
            const marketsForSymbol = _.values(markets).filter(market =>
              market.symbol.startsWith(fsym.toUpperCase())
            );
            for (const market of marketsForSymbol) {
              try {
                const tick = await connector.fetchTicker(market.symbol);
                updatedTicker[market.base] = {
                  ...updatedTicker[market.base],
                  [market.quote]: {
                    PRICE: tick.last || tick.ask,
                    CHANGEPCT24HOUR: 0,
                  },
                };
              } catch (error) {
                console.error(error);
              }
            }
            break;
          }
        }
      }
    }
  } catch (error) {
    console.error(error);
    return ticker;
  }
  return updatedTicker;
};

const exchangeHasSymbol = (symbol: string, exchange: Exchange) => {
  return _.includes(_.keys(exchange.balances), symbol);
};

export function requestHistory(
  fsym: string,
  tsym: string,
  forceRequest: boolean = false
): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const twoMinutesAgo = dayjs().subtract(2, 'minute');
    if (
      forceRequest ||
      !(getHistoryEntry(getHistory(getState()), fsym, tsym).lastUpdate >= twoMinutesAgo.toDate())
    ) {
      fetch(`https://min-api.cryptocompare.com/data/histominute?fsym=${fsym}&tsym=${tsym}`)
        .then(result => result.json())
        .then(result => dispatch(receiveHistory(fsym, tsym, result.Data)))
        .catch(error => console.error(error));
    }
  };
}

export const requestHistoryForAll = (
  fsyms: string[],
  tsym: string,
  forceRequest: boolean = false
) => (dispatch: Dispatch) =>
  fsyms.forEach(fsym => dispatch(requestHistory(fsym, tsym, forceRequest)));

export function requestCoinList(): ThunkAction {
  return (dispatch: Dispatch) => {
    fetch('https://min-api.cryptocompare.com/data/all/coinlist')
      .then(result => result.json())
      .then(result => dispatch(receiveCoinList(result.Data)))
      .catch(error => console.error(error));
  };
}

export function continuouslyUpdateTicker(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    if (!getState().timers.timers.ticker) {
      const timer = window.setInterval(() => dispatch(requestTickerUpdate()), REFRESH_TIME_IN_MS);
      dispatch(startTimer('ticker', timer));
    }
  };
}

interface Symbols {
  from: string[];
  to: string[];
}

function getSymbolsFromTransactions(
  exchanges: Exchanges,
  wallets: Wallet[],
  fiatCurrency: string,
  cryptoCurrency: string,
  extraSymbols: string[]
): Symbols {
  const walletSymbols = wallets.map(wallet => wallet.currency) || [];
  const exchangeSymbols = _.chain(exchanges)
    .values()
    .map(exchange => exchange.balances)
    .map(_.keys)
    .reduce((acc, keys) => acc.concat(keys), [] as string[])
    .value();

  return {
    from: _.uniq([cryptoCurrency, ...walletSymbols, ...exchangeSymbols, ...extraSymbols]),
    to: [cryptoCurrency, fiatCurrency],
  };
}

interface RawTicker {
  [fsym: string]: {
    [tsym: string]: RawTickerEntry;
  };
}

interface RawTickerEntry {
  CHANGEPCT24HOUR: string;
  PRICE: string;
}

function formatRawTicker(rawTicker: RawTicker): Ticker {
  return _.mapValues(rawTicker, fsymEntry =>
    _.mapValues(fsymEntry, (entry: RawTickerEntry) => ({
      ...entry,
      CHANGEPCT24HOUR: parseFloat(entry.CHANGEPCT24HOUR),
      PRICE: parseFloat(entry.PRICE),
    }))
  );
}
