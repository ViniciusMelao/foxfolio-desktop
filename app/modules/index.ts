import { routerReducer as router } from 'react-router-redux';
import { reducer as form } from 'redux-form';
import { persistCombineReducers } from 'redux-persist';
import { configureReduxPersist } from '../store/configureReduxPersist';
import { coinlist} from './coinlist';
import { Coinlist } from './coinlist.types';
import { exchanges } from './exchanges';
import { Exchanges } from './exchanges.types';
import settings from './settings';
import { SettingsType } from './settings.types';
import ticker from './ticker';
import { TickerAndHistory } from './ticker.types';
import timers, { Timers } from './timer';
import { trades} from './trades';
import { TradesMap } from './trades.types';
import { wallets } from './wallets';
import { Wallet } from './wallets.types';

const config = configureReduxPersist();
export const rootReducer = persistCombineReducers(config, {
  coinlist,
  exchanges,
  form,
  router,
  settings,
  ticker,
  timers,
  trades,
  wallets,
} as any);

export interface GlobalState {
  coinlist: Coinlist;
  exchanges: Exchanges;
  settings: SettingsType;
  ticker: TickerAndHistory;
  timers: Timers;
  trades: TradesMap;
  wallets: Wallet[];
}
