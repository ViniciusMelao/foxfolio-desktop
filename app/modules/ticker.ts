import { Action } from '../actions/actions.types';
import { TickerAndHistory } from './ticker.types';

export default function reducer(
  state: TickerAndHistory = { ticker: {}, history: {}, pricesForTime: {} },
  action: Action
): TickerAndHistory {
  switch (action.type) {
    case 'TICKER_UPDATE':
      return { ...state, ticker: action.ticker };
    case 'HISTORY_UPDATE':
      return {
        ...state,
        history: {
          ...state.history,
          [action.fsym]: {
            ...state.history[action.fsym],
            [action.tsym]: {
              lastUpdate: new Date(),
              data: action.history.filter((val, i) => i % 5 === 0),
            },
          },
        },
      };
    case 'RECEIVE_PRICE_FOR_TIME':
      const pricesForSymbol = state.pricesForTime[action.fsym];
      return {
        ...state,
        pricesForTime: {
          ...state.pricesForTime,
          [action.fsym]: {
            ...(state.pricesForTime ? state.pricesForTime[action.fsym] : {}),
            [action.tsym]: {
              ...(pricesForSymbol ? pricesForSymbol[action.tsym] : {}),
              [action.timestamp]: action.price,
            },
          },
        },
      };
    default:
      return state;
  }
}
