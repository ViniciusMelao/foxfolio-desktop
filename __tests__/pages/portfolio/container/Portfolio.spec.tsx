import { createShallow } from '@material-ui/core/test-utils';
import * as React from 'react';
import { Coinlist } from '../../../../app/modules/coinlist.types';
import { initialSettings } from '../../../../app/modules/settings';
import { Ticker } from '../../../../app/modules/ticker.types';
import { EmptyPortfolio } from '../../../../app/pages/portfolio/components/EmptyPortfolio';
import { PortfolioContainer } from '../../../../app/pages/portfolio/container/Portfolio';
import {
  Portfolio,
  PortfolioChange,
  PortfolioSum,
} from '../../../../app/pages/portfolio/types/portfolio.types';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

let coinlist: Coinlist;
let ticker: Ticker;
let portfolio: Portfolio;
let sum: PortfolioSum;
let change: PortfolioChange;

beforeEach(() => {
  coinlist = {};
  ticker = {};
});

test('Render portfolio page', () => {
  ticker = { BTC: { USD: { PRICE: 1000, CHANGEPCT24HOUR: 2 } } };

  portfolio = {
    total: { BTC: 6 },
    exchanges: { exchange: { BTC: 5 } },
    wallets: { BTC: 1 },
  };
  sum = { fiat: 6000, crypto: 6 };

  change = {
    fiat: 6000,
    crypto: 6,
  };

  const wrapper = shallow(
    <PortfolioContainer
      portfolio={portfolio}
      coinlist={coinlist}
      settings={initialSettings}
      ticker={ticker}
      sum={sum}
      change={change}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('Render empty portfolio when ticker is empty', () => {
  portfolio = {
    total: { BTC: 6 },
    exchanges: { exchange: { BTC: 5 } },
    wallets: { BTC: 1 },
  };
  sum = { fiat: 6000, crypto: 6 };

  change = {
    fiat: 6000,
    crypto: 6,
  };

  const wrapper = shallow(
    <PortfolioContainer
      portfolio={portfolio}
      coinlist={coinlist}
      settings={initialSettings}
      ticker={ticker}
      sum={sum}
      change={change}
    />
  );
  expect(wrapper.contains(<EmptyPortfolio />)).toBeTruthy();
});

test('Render empty portfolio when balances and wallets are empty', () => {
  ticker = { BTC: { USD: { PRICE: 1000, CHANGEPCT24HOUR: 2 } } };
  portfolio = {
    total: {},
    exchanges: {},
    wallets: {},
  };
  sum = { fiat: 0, crypto: 0 };

  change = {
    fiat: 0,
    crypto: 0,
  };
  const wrapper = shallow(
    <PortfolioContainer
      portfolio={portfolio}
      coinlist={coinlist}
      settings={initialSettings}
      ticker={ticker}
      sum={sum}
      change={change}
    />
  );
  expect(wrapper.contains(<EmptyPortfolio />)).toBeTruthy();
});
