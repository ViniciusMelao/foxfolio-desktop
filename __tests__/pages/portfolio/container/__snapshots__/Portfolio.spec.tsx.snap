// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render portfolio page 1`] = `
<div>
  <WithStyles(Paper)
    style={
      Object {
        "marginTop": 0,
        "paddingBottom": 25,
        "paddingTop": 25,
        "textAlign": "center",
      }
    }
  >
    <Component
      change={
        Object {
          "crypto": 6,
          "fiat": 6000,
        }
      }
      settings={
        Object {
          "cryptoCurrency": "BTC",
          "currencyFocus": "fiat",
          "fiatCurrency": "USD",
          "hideZeroBalances": false,
          "includeFiat": true,
          "theme": "light",
        }
      }
      sum={
        Object {
          "crypto": 6,
          "fiat": 6000,
        }
      }
    />
  </WithStyles(Paper)>
  <WithStyles(Paper)
    style={
      Object {
        "marginTop": 30,
        "paddingBottom": 20,
        "paddingTop": 10,
      }
    }
  >
    <WithTheme(Component)
      portfolio={
        Object {
          "BTC": 6,
        }
      }
      settings={
        Object {
          "cryptoCurrency": "BTC",
          "currencyFocus": "fiat",
          "fiatCurrency": "USD",
          "hideZeroBalances": false,
          "includeFiat": true,
          "theme": "light",
        }
      }
      sum={6}
      ticker={
        Object {
          "BTC": Object {
            "USD": Object {
              "CHANGEPCT24HOUR": 2,
              "PRICE": 1000,
            },
          },
        }
      }
    />
  </WithStyles(Paper)>
  <div
    style={
      Object {
        "marginTop": 30,
      }
    }
  >
    <WithStyles(Paper)
      square={true}
      style={
        Object {
          "width": 192,
        }
      }
    >
      <Connect(DisconnectedPortfolioFocusButtons) />
    </WithStyles(Paper)>
  </div>
  <div>
    <Component
      coinlist={Object {}}
      portfolio={
        Object {
          "exchanges": Object {
            "exchange": Object {
              "BTC": 5,
            },
          },
          "total": Object {
            "BTC": 6,
          },
          "wallets": Object {
            "BTC": 1,
          },
        }
      }
      settings={
        Object {
          "cryptoCurrency": "BTC",
          "currencyFocus": "fiat",
          "fiatCurrency": "USD",
          "hideZeroBalances": false,
          "includeFiat": true,
          "theme": "light",
        }
      }
      ticker={
        Object {
          "BTC": Object {
            "USD": Object {
              "CHANGEPCT24HOUR": 2,
              "PRICE": 1000,
            },
          },
        }
      }
    />
  </div>
</div>
`;
